src/Tools/dotnet-dsrouter/DiagnosticsServerRouterCommands.cs

```diff
@@ -45,254 +45,290 @@ public void OnRouterStopped()
-
-            LogLevel logLevel = LogLevel.Information;
-            if (string.Compare(verbose, "debug", StringComparison.OrdinalIgnoreCase) == 0)
-                logLevel = LogLevel.Debug;
-            else if (string.Compare(verbose, "trace", StringComparison.OrdinalIgnoreCase) == 0)
-                logLevel = LogLevel.Trace;
+            var runner = new IpcClientTcpServerRunner(verbose);
 
-            using var factory = new LoggerFactory();
-            factory.AddConsole(logLevel, false);
+            return await runner.CommonRunLoop((logger, launcherCallbacks, linkedCancelToken) =>
+            {
+                NetServerRouterFactory.CreateInstanceDelegate tcpServerRouterFactory = ChooseTcpServerRouterFactory(forwardPort, logger);
 
-            Launcher.SuspendProcess = false;
-            Launcher.ConnectMode = true;
-            Launcher.Verbose = logLevel != LogLevel.Information;
-            Launcher.CommandToken = token;
+                var routerTask = DiagnosticsServerRouterRunner.runIpcClientTcpServerRouter(linkedCancelToken.Token, ipcClient, tcpServer, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, tcpServerRouterFactory, logger, launcherCallbacks);
+                return routerTask;
+            }, token);
+        }
 
-            var logger = factory.CreateLogger("dotnet-dsrouter");
+        class IpcServerTcpServerRunner : SpecificRunnerBase
+        {
+            public IpcServerTcpServerRunner(string verbose) : base(verbose) { }
 
-            TcpServerRouterFactory.CreateInstanceDelegate tcpServerRouterFactory = TcpServerRouterFactory.CreateDefaultInstance;
-            if (!string.IsNullOrEmpty(forwardPort))
+            public override void ConfigureLauncher(CancellationToken cancellationToken)
             {
-                if (string.Compare(forwardPort, "android", StringComparison.OrdinalIgnoreCase) == 0)
-                {
-                    tcpServerRouterFactory = ADBTcpServerRouterFactory.CreateADBInstance;
-                }
-                else
-                {
-                    logger.LogError($"Unknown port forwarding argument, {forwardPort}. Only Android port fowarding is supported for TcpServer mode. Ignoring --forward-port argument.");
-                }
+                Launcher.SuspendProcess = false;
+                Launcher.ConnectMode = true;
+                Launcher.Verbose = LogLevel != LogLevel.Information;
+                Launcher.CommandToken = cancellationToken;
             }
+        }
 
-            if (string.IsNullOrEmpty(ipcServer))
-                ipcServer = GetDefaultIpcServerPath(logger);
+        public async Task<int> RunIpcServerTcpServerRouter(CancellationToken token, string ipcServer, string tcpServer, int runtimeTimeout, string verbose, string forwardPort)
+        {
+            checkLoopbackOnly(tcpServer);
```

Comment:
When setting up TcpServer we have this warning printed if you try to set it up not using loopback interface (since you expose a port that is unautenticated and unencrypted otherwise). Maybe we should have something similar if you try to set the web server without TLS? Is it possible to setup the web socket server (and the client running in web assembly) to use TLS for the encryption channel?

Reply:
The right place to warn would be in the .NET runtime WASM SDK, I think - when the DiagnosticPorts environment variable is set for the project.